azure-core/azure/core/pipeline/policies/universal.py

```diff
@@ -0,0 +1,323 @@
+        self._headers.update(kwargs.pop('headers', {}))
+
+    @property
+    def headers(self):
+        """The current headers collection."""
+        return self._headers
+
+    def add_header(self, key, value):
+        """Add a header to the configuration to be applied to all requests."""
+        self._headers[key] = value
+
+    def on_request(self, request):
+        # type: (PipelineRequest, Any) -> None
+        request.http_request.headers.update(self.headers)
+        additional_headers = request.context.options.pop('headers', {})
+        if additional_headers:
+            request.http_request.headers.update(additional_headers)
+
+
+class UserAgentPolicy(SansIOHTTPPolicy):
+    _USERAGENT = "User-Agent"
+    _ENV_ADDITIONAL_USER_AGENT = 'AZURE_HTTP_USER_AGENT'
+
+    def __init__(self, base_user_agent=None, **kwargs):
+        # type: (Optional[str], bool) -> None
+        self.overwrite = kwargs.pop('user_agent_overwrite', False)
+        self.use_env = kwargs.pop('user_agent_use_env', True)
+
+        if base_user_agent is None:
+            self._user_agent = "python/{} ({}) azure-core/{}".format(
+                platform.python_version(),
+                platform.platform(),
+                azcore_version
+            )
+        else:
+            self._user_agent = base_user_agent
+
+    @property
+    def user_agent(self):
+        # type: () -> str
+        """The current user agent value."""
+        if self.use_env:
+            add_user_agent_header = os.environ.get(self._ENV_ADDITIONAL_USER_AGENT, None)
+            if add_user_agent_header is not None:
+                return "{} {}".format(self._user_agent, add_user_agent_header)
+        return self._user_agent
+
+    def add_user_agent(self, value):
+        # type: (str) -> None
+        """Add value to current user agent with a space.
+
+        :param str value: value to add to user agent.
+        """
+        self._user_agent = "{} {}".format(self._user_agent, value)
+
+    def on_request(self, request):
+        # type: (PipelineRequest, Any) -> None
+        http_request = request.http_request
+        options = request.context.options
+        if 'user_agent' in options:
+            user_agent = options.pop('user_agent')
+            if options.pop('user_agent_overwrite', self.overwrite):
+                http_request.headers[self._USERAGENT] = user_agent
+            else:
+                user_agent = "{} {}".format(self.user_agent, user_agent)
+                http_request.headers[self._USERAGENT] = user_agent
+
+        elif self.overwrite or self._USERAGENT not in http_request.headers:
+            http_request.headers[self._USERAGENT] = self.user_agent
+
+
+class NetworkTraceLoggingPolicy(SansIOHTTPPolicy):
+    """A policy that logs HTTP request and response to the DEBUG logger.
+
+    This accepts both global configuration, and per-request level with "enable_http_logger"
+    """
+    def __init__(self, logging_enable=False):
```

Comment:
Should this accept kwargs like other policies? (autorest assumes it does)

Reply:
Yeah can't hurt.